HDR Output support

FrancescoC-unity opened the pull request:
_NOTE:_ PR has lots of files that have lots of actual changes compressed by github, please when reviewing expand the diff. 

This PR introduces proper HDR Output support for HDRP. 

TL;DR of the features and characteristics: 

- Support for HDR10 and scRGB output standard. 
- ACES workflow for HDR (3 presets, 1000, 2000 and 4000 nits)
- Adding Full ACES even for LDR workflow.
- Neutral tonemapper using standard BT2390 curve for range reduction.
- Added control over how much hue shift is desired. 
- Control over target min/max nits and display paper white. 
- Proper UI handling so that white in UI is maintained as white.
- Added several debug visualizations, with special focus on gamut visualization. 
- Changed the whole grading pipeline to always use ACEScg as working space and allowing to go back to sRGB. 
- Optimized LUT building so it is created only when a change happens. 

When enabling HDR all the HDR Output tunable parameters are shown in the Tonemapping volume component
![image](https://user-images.githubusercontent.com/43168857/136534182-8e3c41bb-04dd-41a1-a1c4-f121747b23aa.png)
![image](https://user-images.githubusercontent.com/43168857/136534396-81bc8bd1-703f-478e-b87a-5292abd48432.png)


Unfortunately cannot post images as they wouldn't show much on LDR screens and LDR screenshots :D The debug view can be seen here though

![image](https://user-images.githubusercontent.com/43168857/136533547-cc3192de-d2ef-4a86-8d7e-6b3883433bfe.png)
![image](https://user-images.githubusercontent.com/43168857/136533587-a1dd3e0e-7a36-4fbf-920b-d3e8beffe0d6.png)
![image](https://user-images.githubusercontent.com/43168857/136533606-e3679332-b767-4d7e-a9ae-37f5cec02979.png)


Most of the math for conversion and tone curves are in [HDROutput.hlsl ](https://github.com/Unity-Technologies/Graphics/blob/a3a7d7fb76b1914922104e6f887a626a0a366ae5/com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl) I tried to be verbose with comments there and added a bunch of desmos 


A lot more can be written, but I am keeping this PR description tight. Please reach out and I am happy to discuss any aspect in more details. 


I will be start writing docs now and will probably open a separate doc PR when I am done. Though I also need to find a way to fake a LDR/HDR comparison in LDR for Docs, and I have no clue how to do that :D

FrancescoC-unity commented:
Given the change in color grading I am expecting a few failed tests, I am running yamato and fix the failing ones.

JulienIgnace-Unity commented on com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.PostProcess.cs:
```diff
@@ -284,12 +289,14 @@ void CleanupPostProcess()
             CoreUtils.Destroy(m_ClearBlackMaterial);
             CoreUtils.Destroy(m_SMAAMaterial);
             CoreUtils.Destroy(m_TemporalAAMaterial);
+            CoreUtils.Destroy(m_GradingAndTonemappingLUT);
```
Should be RTHandles.Release here (or m_GradingAndTonemappingLUT.Release)

FrancescoC-unity commented:
> Please update What's new and upgrade guide (or do it in the doc PR if it is not done there :) ) (in particular for this: Changed the whole grading pipeline to always use ACEScg as working space and allowing to go back to sRGB.)

Added it in the doc PR: https://github.com/Unity-Technologies/Graphics/pull/5992/files#diff-e64740d8eca588a80a48a41a51c9f66e6b56062814a5b273b4b399cae0a5a689R3

sebastienlagarde commented on com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl:
```diff
@@ -501,16 +507,19 @@ half segmented_spline_c5_fwd(half x)
+    float slopeHigh;      // log-log slope of high linear extension
+};
 
+half segmented_spline_c9_fwd(half x, SegmentedSplineParams_c9 params)
```
Note: we will need to mention this change in upgrade guide

FrancescoC-unity replied:
Yeah to be safe I added the old signature/version so it is compatible and should work like before

sebastienlagarde commented on com.unity.render-pipelines.core/ShaderLibrary/ACES.hlsl:
```diff
@@ -762,14 +771,27 @@ static const half ODT_SAT_FACTOR = 0.93;
+    const SegmentedSplineParams_c9 ODT_48nits =
+    {
+        // coefsLow[10]
+        { -1.6989700043, -1.6989700043, -1.4779000000, -1.2291000000, -0.8648000000, -0.4480000000, 0.0051800000, 0.4511080334, 0.9113744414, 0.9113744414},
```
also if I am not wrong this code is duplicated few time, can't we declare it a single time?

FrancescoC-unity replied:
Originally I had it statically defined, but it generated obscene code, I can have a function returning the factors though, will do.

sebastienlagarde commented on com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:
```diff
@@ -0,0 +1,657 @@
+
+float3 RotateOutputSpaceToICtCp(float3 inputColor)
+{
+    // TODO: Do the conversion directly from Rec709 (bake matrix Rec709 -> XYZ -> LMS)
```
guess the compiler may do it?

FrancescoC-unity replied:
This is a bit different, what I mean is essentially have a version that is RotateRec2020ToICtCp without needing the extra step in between. 

I am not sure the compiler will be able to merge matrices by itself. 

That said, this is not a big deal as it all happen during LUT building.

sebastienlagarde commented on com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:
```diff
@@ -0,0 +1,657 @@
+#define ISS_APPROX_PQ 1
+#define GTS_APPROX_PQ 2
+
+#define OETF_CHOICE GTS_APPROX_PQ
```
is it worse it? the PRECISE_PQ dont look like so expensive. Or maybe change depends on platform? approx on console but full on PC?

FrancescoC-unity replied:
It is still a few trascendentals, the approx is very very similar (see the desmos https://www.desmos.com/calculator/5jdfc4pgtk )

FrancescoC-unity replied:
Check the error diff in the diff, it is very very minimal (bottom of the desmos)

sebastienlagarde commented on com.unity.render-pipelines.core/ShaderLibrary/HDROutput.hlsl:
```diff
@@ -0,0 +1,657 @@
+}
+
+// Ref "High Dynamic Rangecolor grading and displayin Frostbite" [Fry 2017]
+float3 FryHuePreserving(float3 input, float minNits, float maxNits, float hueShift, int mode)
```
not used anywhere?

FrancescoC-unity replied:
No, but I'd like to keep as I expect some feedback at some point about this hue preservation. I rolled our own for the same thing, but if we need to revisit I'd like to have this here.

sebastienlagarde commented on com.unity.render-pipelines.high-definition/Editor/PostProcessing/TonemappingEditor.cs:
```diff
@@ -125,6 +152,57 @@ public override void OnInspectorGUI()
+                int hdrTonemapMode = m_Mode.value.intValue;
+                if (m_Mode.value.intValue == (int)TonemappingMode.Custom || hdrTonemapMode == (int)TonemappingMode.External)
+                {
+                    EditorGUILayout.HelpBox("The selected tonmapping mode is not supported in HDR Output mode. Select a fallback mode.", MessageType.Warning);
```
typo tonmapping

sebastienlagarde commented on com.unity.render-pipelines.high-definition/Runtime/Debug/DebugHDR.shader:
```diff
@@ -0,0 +1,333 @@
+        return float4(color * lineAlpha, lineAlpha);
+    }
+
+    float3 xyYtoXYZ(float3 xyY)
```
don't we have those function in color.hlsl? we should move them there

sebastienlagarde replied:
ok

sebastienlagarde commented on com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Components/Tonemapping.cs:
```diff
@@ -20,7 +20,7 @@ public enum TonemappingMode
 
         /// <summary>
-        /// Close approximation of the reference ACES tonemapper for a more filmic look.
+        /// ACES tonemapper for a more filmic look.
```
to clarify, it mean we don't have approx anymore for ACES? (guess you told me but I forget), does it mean we have a perf hit here?

FrancescoC-unity replied:
We have both now, approx or not depends on a checkbox (scroll a bit down the file for the checkbox)

FrancescoC-unity replied:
By default the checkbox is off, no need for anything to do when upgrading. Isn't it more a what's new?

FrancescoC-unity commented on com.unity.render-pipelines.high-definition/Runtime/PostProcessing/Shaders/DebugHDRxyMapping.compute:
```diff
@@ -0,0 +1,43 @@
+#define _SizePerDim _HDRxyBufferDebugParams.xy
+#define _IsRec709 (int)(_HDRxyBufferDebugParams.w == 1)
+
+float2 RGBtoxy(float3 rgb)
```
Moved at least partially
